### ISSUE TITLE
[Technical Debt] Remove or Refactor Simplified Forward Pass Implementation in Production Codebase

### ISSUE THREAD

**Issue by EffortlessSteven** [EffortlessMetrics/BitNet-rs#352]
# [Technical Debt] Remove or Refactor Simplified Forward Pass Implementation in Production Codebase

## Problem Description

The `simple_forward.rs` module in `crates/bitnet-inference/src/` provides a minimal forward pass implementation that bypasses the full transformer architecture, implementing only token embedding lookup and language model head multiplication. This simplified implementation poses significant risks to production neural network inference validation and may mislead developers about actual model behavior.

## Environment

- **File**: `crates/bitnet-inference/src/simple_forward.rs`
- **Module**: `bitnet-inference::simple_forward`
- **Public API**: `Weights` struct and `logits_for_token()` function
- **Current Usage**: `examples/infer.rs` (demonstration purposes)
- **MSRV**: Rust 1.90.0
- **Features**: Available in all builds (no feature gating)

### Technical Investigation

The `simple_forward.rs` module implements a forward pass that:

1. **Skips Critical Transformer Components**: Bypasses attention mechanisms, layer normalization, feed-forward networks, and all intermediate transformer layers
2. **Provides False Validation**: Only tests embedding lookup (`tok_embeddings[token_id]`) and matrix multiplication (`e @ lm_head`)
3. **Lacks Production Context**: Does not represent real neural network inference behavior
4. **Creates Validation Blind Spots**: May pass validation while actual transformer layers contain bugs

### Architecture Concerns

```rust
// Current simplified implementation - bypasses entire transformer
pub fn logits_for_token(w: &Weights, token_id: usize) -> Vec<f32> {
    // e = tok_embeddings[token_id]  // [dim]
    let e = &w.tok_embeddings[token_id * w.dim..(token_id + 1) * w.dim];

    // out = e @ lm_head  // [vocab] - SKIPS ALL TRANSFORMER LAYERS
    // Missing: attention, layer norm, feed-forward, positional encoding, etc.
}
```

### Usage Analysis

**Current Usage Pattern:**
- **Primary**: `examples/infer.rs` for demonstration
- **Exposure**: Public API in `bitnet-inference::simple_forward` module
- **Risk**: Available for import in production code without clear warnings

**Missing Safeguards:**
- No `#[cfg(test)]` or `#[cfg(feature = "examples")]` gates
- No documentation warnings about production usage
- No clear separation from real inference APIs

### Severity: **Medium-High**

**Affected Components:**
- Neural network inference validation
- Developer testing workflows
- API surface clarity
- Production readiness assessment

**Potential Risks:**
1. **False Confidence**: Tests using this may pass while real inference fails
2. **Performance Misrepresentation**: Simplified path doesn't reflect actual computational costs
3. **Debugging Confusion**: May mask issues in actual transformer implementation
4. **API Pollution**: Unclear public interface between test and production code

**Business Impact:**
- Could lead to undetected inference bugs reaching production
- Reduces confidence in validation framework
- Increases technical debt and maintenance burden

### Primary Approach: Restrict to Test/Example Context

1. **Feature Gate Implementation**:
   ```rust
   // Option A: Test-only gating
   #[cfg(any(test, feature = "examples"))]
   pub mod simple_forward;

   // Option B: Explicit example feature
   #[cfg(feature = "examples")]
   pub mod simple_forward;
   ```

2. **Add Clear Documentation Warnings**:
   ```rust
   //! # ⚠️ WARNING: Test/Example Only
   //!
   //! This module provides a **simplified forward pass for testing only**.
   //! It bypasses all transformer layers (attention, layer norm, feed-forward)
   //! and should **never be used for production inference validation**.
   ```

3. **Rename for Clarity**:
   ```rust
   // Current: simple_forward.rs
   // Proposed: test_forward.rs or example_forward.rs
   ```

### Alternative Approaches

**Option B: Remove Entirely**
- Delete the module and update `examples/infer.rs` to use real inference
- Pros: Eliminates confusion, forces proper testing
- Cons: May break existing development workflows

**Option C: Convert to Benchmarking Tool**
- Move to `benches/` or dedicated performance testing module
- Use for measuring embedding/projection performance only
- Clear separation from inference validation

### Phase 1: Immediate Safety (1-2 days)
- [ ] Add feature gate: `#[cfg(feature = "examples")]`
- [ ] Update `examples/infer.rs` to require `--features examples`
- [ ] Add prominent documentation warnings
- [ ] Update `Cargo.toml` to include `examples` feature

### Phase 2: API Cleanup (2-3 days)
- [ ] Rename module to `example_forward` or `test_forward`
- [ ] Update imports in affected files
- [ ] Ensure no production code dependencies
- [ ] Add integration test to verify feature gating works

### Phase 3: Documentation & Validation (1 day)
- [ ] Update architecture documentation
- [ ] Add section to `docs/development/` explaining test vs production inference
- [ ] Create clear migration guide for any external users

### Validation Requirements
1. **Feature Gate Verification**:
   ```bash
   # Should fail without examples feature
   cargo build --no-default-features --features cpu
   cargo test --workspace --no-default-features --features cpu

   # Should succeed with examples feature
   cargo build --no-default-features --features "cpu,examples"
   cargo run --example infer --features "cpu,examples"
   ```

2. **Integration Testing**:
   - Verify no production code imports `simple_forward`
   - Confirm examples work with proper feature flags
   - Test that documentation clearly separates concerns

3. **Regression Testing**:
   - Ensure real inference paths still function
   - Verify no performance impact on production builds
   - Confirm API compatibility for legitimate test usage

## Acceptance Criteria

- [ ] **Safety**: `simple_forward` module only available with explicit feature flag
- [ ] **Clarity**: Documentation clearly warns against production usage
- [ ] **Isolation**: No production code can accidentally import simplified implementation
- [ ] **Functionality**: Examples continue to work with proper feature configuration
- [ ] **Testing**: Integration tests verify feature gating behavior
- [ ] **Documentation**: Clear guidance on test vs production inference approaches

## Related Issues

This issue is part of a broader effort to eliminate simulation/mock implementations:
- [TBD] Mock object consolidation in backends and engines
- [TBD] GPU memory management simulation removal
- [TBD] Validation framework improvements for production readiness

## Technical References

- **Module**: `crates/bitnet-inference/src/simple_forward.rs`
- **Usage**: `examples/infer.rs:44-52`
- **Public API**: `bitnet-inference::simple_forward::{Weights, logits_for_token}`
- **Architecture docs**: `docs/architecture-overview.md`
- **Feature documentation**: `docs/explanation/FEATURES.md`
